core/dive.c

```diff
@@ -795,10 +795,19 @@ pressure_t calculate_surface_pressure(const struct dive *dive)
 	pressure_t res;
 	int sum = 0, nr = 0;
 
-	for_each_dc (dive, dc) {
-		if (dc->surface_pressure.mbar) {
-			sum += dc->surface_pressure.mbar;
-			nr++;
+	if (has_planned(dive, false)) {
```

Comment:
I'm sure I'm missing something obvious, so please bear with me... but...
why the if-clause... why not simply do a `for_each_real_dc()` ??

Reply:
Otherwise, a dive that has only planned dive computers would end up without maxdepth, total time etc.

Reply:
Or said differently: This PR should only influence dives that have both: real and planned dive computers.

Reply:
but then... wouldn't it make more sense to have the macro do the right thing? i.e., move that logic into the macro.
If there are only plans, loop over the plans, otherwise, only loop over real dive computers? I just dislike the fact that we have the identical code twice, and that we do need this weird construct multiple times. Seems more error prone.
So I guess I'm suggesting something like
```for_each_relevant_dc()```

Reply:
Yes, I though about this. But I could not figure out how to do this in a macro (include the if statement only in case some condition holds). My preprocessor experience is even smaller than my C experience (I was already quite proud of my new macro). Of course, one could include the first if with and in the condition of the inner if but then this condition would be tested for every dive computer (and the condition even loops over all dive computers as well). Could you suggest a macro that does this? If you can come up with something reasonable, I would be more than happy to use that instead.

Reply:
Well, the simplest way to do this would be something like this:
```
#define for_each_relevant_dc(_dive, _dc) \
	bool _all_planned = !has_planned(_dive, false); \
	for (_dc = &_dive->dc; _dc; _dc = _dc->next) if (_all_planned || !is_dc_planner(_dc))
```
that does of course leak the bool into the calling name space, which means that you CANNOT use it twice in the same block... but it seems straight forward enough.
Complexity wise, this is the same as your code - it's just easier on my eyes...

Now, if you would like a version that you can use more than once in same block, it gets a bit (ummm...) uglier:
```
// you need the secondary macro invocation in order for __LINE__ to be expanded so the concatenation works
// this way the for_each_relevant_dc macro can be used multiple times within the same name space
#define _UNIQ_NAME(_a) _CONCAT(_a, __LINE__)
#define _CONCAT(_a, _l) _CONCAT_HACK(_a, _l)
#define _CONCAT_HACK(_a, _l) _a ## _l
#define for_each_relevant_dc(_dive, _dc) \
	bool _UNIQ_NAME(_all_planned) = !has_planned(_dive, false); \
	for (_dc = &_dive->dc; _dc; _dc = _dc->next) if (_UNIQ_NAME(_all_planned) || !is_dc_planner(_dc))

```
(and for the people who are pain-curious... yes, the `__LINE__` macro expands to the line number in which `for_each_relevant_dc()` gets called, so this does work, even though it looks mighty unintuitive...)

Reply:
Thanks, I used the first version (if somebody has to use it twice in a row, throw in an extra pair of braces and it will still work. Indeed, this makes it much nicer.

Reply:
I'm glad you like it!